File: MonkeyTrap/app/src/main/java/com/novoda/monkeytrap/MainActivity.java
package com.novoda.monkeytrap;

import android.app.Activity;
import android.os.Bundle;
import android.widget.CompoundButton;
import android.widget.Switch;

public class MainActivity extends Activity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        Switch overlaySwitch = (Switch) findViewById(R.id.overlay_switch);
        overlaySwitch.setChecked(OverlayService.isOverlayShowing());
        overlaySwitch.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                if (isChecked) {
                    startService(OverlayService.show());
                } else {
                    startService(OverlayService.hide());
                }
            }
        });
    }

}

File: sliceanddice/host/src/main/java/com/novoda/sliceanddice/MainActivity.kt
package com.novoda.sliceanddice

import android.net.Uri
import android.os.Bundle
import android.widget.ArrayAdapter
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.slice.SliceManager
import androidx.slice.widget.SliceLiveData
import com.novoda.sliceshost.R
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.settings_card.*

class MainActivity : AppCompatActivity() {

    private lateinit var sliceManager: SliceManager

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        sliceManager = SliceManager.getInstance(this)
        sliceSelector.adapter = ArrayAdapter<SliceChoice>(this, android.R.layout.simple_list_item_1, sliceChoices)
    }

    override fun onStart() {
        super.onStart()

        if (providerAppNotInstalled(packageManager, baseSliceUri.authority)) {
            showMissingProviderDialog(this, { finish() }, baseSliceUri)
            return
        }

        sliceSelector.onItemSelectedListener = sliceSelectionListener({ _, _, selectedPosition, _ ->
            onSliceSelectionChanged(sliceChoices[selectedPosition].uri)
        })

        sliceSelector.setSelection(0)
    }

    private fun onSliceSelectionChanged(selectedSliceUri: Uri) {
        startObservingSliceLiveData(selectedSliceUri)
    }

    private fun startObservingSliceLiveData(sliceUri: Uri) {
        val slice = sliceManager.bindSlice(sliceUri)
        sliceView.slice = slice

        SliceLiveData.fromUri(this, sliceUri)
            .observe(this, Observer({ sliceResult ->
                sliceView.slice = sliceResult
                invalidateOptionsMenu()
            }))
    }
}
